packages/react/ds-app-launchpad/src/ui/MarkdownEditor/MarkdownEditor.tsx

```diff
@@ -0,0 +1,188 @@
+/* @canonical/generator-ds 0.9.0-experimental.4 */
+import hljs from "highlight.js";
+import type React from "react";
+import {
+  forwardRef,
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useMemo,
+  useRef,
+  useState,
+} from "react";
+import ReactMarkdown from "react-markdown";
+import rehypeSanitize from "rehype-sanitize";
+import "../GitDiffViewer/common/CodeDiffViewer/HighlighTheme.css";
+import {
+  Toolbar,
+  ToolbarButton,
+  ToolbarGroup,
+  ToolbarSeparator,
+  ViewModeTabs,
+  icons,
+} from "./common/index.js";
+import useEditor from "./hooks/useEditor.js";
+import "./styles.css";
+import type { EditMode, MarkdownEditorProps } from "./types.js";
+
+const componentCssClassName = "ds markdown-editor";
+
+/**
+ * A dual-mode Markdown editor for writing and previewing Markdown content.
+ */
+const MarkdownEditor = (
+  {
+    id,
+    className,
+    style,
+    defaultValue,
+    placeholder,
+    hideToolbar = false,
+    hidePreview = false,
+    editMode: controlledEditMode,
+    onEditModeChange: controlledOnEditModeChange,
+  }: MarkdownEditorProps,
+  ref: React.Ref<HTMLTextAreaElement | null>,
+): React.ReactElement => {
+  const textareaRef = useRef<HTMLTextAreaElement>(null);
+  const previewRef = useRef<HTMLDivElement>(null);
+  const [internalEditMode, setInternalEditMode] = useState<EditMode>("write");
+  const { toolbar } = useEditor(textareaRef);
+
+  const editMode = useMemo(() => {
+    return controlledEditMode ?? internalEditMode;
+  }, [controlledEditMode, internalEditMode]);
+
+  const handleEditModeChange = useCallback(
```

Comment:
When switching from `Preview` to `Edit`, it would be nice if the TextEntry were automatically focused.

Reply:
Sure! Will add the auto focus on click event.

Reply:
Thanks, this works great!